sematic/log_reader.py

```diff
@@ -0,0 +1,429 @@
+    filter_strings:
+        Only log lines that contain ALL of the strings in this list will
+        be included in the result
+
+    Returns
+    -------
+    A subset of the logs for the given run
+    """
+    run = get_run(run_id)
+    run_state = FutureState[run.future_state]  # type: ignore
+    still_running = run_state.is_terminal() or run_state == FutureState.RAN
+    resolution = get_resolution(run.root_id)
+    filter_strings = filter_strings if filter_strings is not None else []
+    cursor = (
+        Cursor.from_token(continuation_cursor)
+        if continuation_cursor is not None
+        else Cursor.nothing_found(filter_strings, run_id)
+    )
+    if cursor.run_id != run_id:
+        raise ValueError(
+            f"Tried to continue a log search of {run_id} using a "
+            f"continuation cursor from {cursor.run_id}"
+        )
+    if set(cursor.filter_strings) != set(filter_strings):
+        raise ValueError(
+            f"Tried to continue a log search of {run_id} using a "
+            f"different set of filters than were used in the cursor."
+        )
+
+    if ResolutionStatus[resolution.status] in (  # type: ignore
+        ResolutionStatus.CREATED,
+        ResolutionStatus.SCHEDULED,
+    ):
+        return LogLineResult(
+            more_before=False,
+            more_after=True,
+            lines=[],
+            continuation_cursor=cursor.to_token(),
+            log_unavaiable_reason="Resolution has not started yet.",
+        )
+    filter_strings = filter_strings if filter_strings is not None else []
+    if FutureState[run.future_state] == FutureState.CREATED:  # type: ignore
+        return LogLineResult(
+            more_before=False,
+            more_after=True,
+            lines=[],
+            continuation_cursor=cursor.to_token(),
+            log_unavaiable_reason="The run has not yet started executing.",
+        )
+    # looking for external jobs to determine inline is only valid
+    # since we know the run has at least reached SCHEDULED due to it
+    # not being CREATED.
+    is_inline = len(run.external_jobs) == 0
+    if is_inline:
+        return _load_inline_logs(
+            run_id=run_id,
+            resolution=resolution,
+            still_running=still_running,
+            cursor_file=cursor.source_log_key,
+            cursor_line_index=cursor.source_file_line_index,
+            max_lines=max_lines,
+            filter_strings=filter_strings,
+        )
+    return _load_non_inline_logs(
+        run_id=run_id,
+        still_running=still_running,
+        cursor_file=cursor.source_log_key,
+        cursor_line_index=cursor.source_file_line_index,
+        max_lines=max_lines,
+        filter_strings=filter_strings,
+    )
+
+
+def _get_latest_log_file(prefix, cursor_file) -> Optional[str]:
+    log_files = storage.get_child_paths(prefix)
+    if len(log_files) < 1:
+        return None
+
+    # the file wth the highest timestamp has the full logs.
```

Comment:
I don't understand this. For a given run ID, do we duplicate logs across multiple files?

Reply:
Right now, yes. We upload to s3 a snapshot of the *entire* logs at various timestamps. So file 1 may contain lines 0-10, file 2 may contain 0-20, etc.. until file n contains lines 0-N. This is less efficient than having each file store a separate portion of the logs, but there's not a great way to do that without risking either (a) duplicating lines (b) omitting lines (c) complicating ingestion with some fancy mechanisms around making sure we don't do either of the above.

Reply:
The easiest way to avoid those two problems would be to use a pipe-based rather than file-based ingestion, and the "sidecar process" can just chunk things up into groups of lines as they stream in from the main process. There's a ticket to do that later.

Reply:
Ok so we have a strategy to fix this? Is the current setup forward compatible with the future setup?

Reply:
Yes and kind-of. The strategy involves having some period of time where both structures of log file are supported. So users wouldn’t lose access to old logs, but we would need to keep some code around for reading them for a few versions.

Reply:
here's the ticket, FWIW: #155